### ISSUE TITLE
Support states which are non default constructable

### ISSUE THREAD

**Issue by erikzenker** [erikzenker/hsm#102]
**Problem:**
States need to be default constructable. Thus you can't use a constructor to initialize state data members.

**Current workaround:**
Initialize a state in its entry action by setting its members or calling an init function
```c++
struct Target {
    constexpr auto on_entry()
    {
        return [](auto event, auto source, auto& target) { 
          target.member = std::string("init String"); 
       };
    }
};
```

**Proposed solution by** @BenjaminW3:
 provide a way to use the constructor of a class and every action call would create a new instance of the target class such as:

```c++
const auto action = [](auto&& event, auto source) {
  return Target{std::string("Init string")};
};
```

or return an instance which is saved externally:

```c++
const auto action = [](auto&& event, auto source, auto& dependency) {
  return dependency.target;
}
```

**Comment by erikzenker** [comment]
I am thinking about the problem and possible solutions. I don't think that returning the target object in the current actions is the most sophisticated solution to initialize the state. It will make it the transition table harder to understand. Two other solutions that came into my mind:

1. Provide state objects to the state machine object.
    The problem here is that you might not have all parameters to construct the state yet.

2. Provide a special on_entry event as it is done in the [sml](https://boost-ext.github.io/sml/examples.html#states)
    The action of this event could return the constructed target state
    ```c++
   hsm::state<S1>{} + hsm::on_entry{} / [](auto event, auto source){ return S1{std::string{"Init string"}};}  
    ```

**Comment by erikzenker** [comment]
@BenjaminW3 can you give me more details on your use case?

**Comment by BenjaminW3** [comment]
In the use case I have in mind, the target state is constructed based on the members of the event leading to the transition to the to-be-constructed target state.
Using `on_entry` is probably already too late because there the event is already gone.

**Comment by BenjaminW3** [comment]
For reference the rest of my proposal from the other PR discussion:
"I would still allow void returns from actions to use the default constructor. Furthermore, we would have to make sure that only one of the actions in a transition (maybe only the last one?) may return the next state"

**Comment by erikzenker** [comment]
Just for the record: I found two bugs while writing the on_entry state initialization integration test (see #103 and #104)

**Comment by erikzenker** [comment]
#103 and #104 were fixed

**Comment by erikzenker** [comment]
Accidentally, I found out that the sml also supports [data members](https://github.com/boost-ext/sml/blob/master/example/data.cpp). Maybe because it is not mentioned in the documentation I missed the feature until now.

Just for the record, the user can provide non default initialized states to the [constructor](https://github.com/boost-ext/sml/blob/master/example/data.cpp#L63) of the state machine. I guess if he doesn't they will be default initialized.

The solution is like the first one I already discussed above. And it still has the problem that some states might need data from an event to be constructed.
